Make transforms stateless

brandonwillard opened the pull request:
This PR addresses a few more transform changes/issues.

The primary change is that transforms are now stateless (i.e. they no longer carry their own parameters).   Stateful transforms make it very easy to accidentally introduce old and/or irrelevant parameters into a graph, and are a source for some extremely confusing and difficult bugs.  That's why this change was made.

Now, transforms only take a "parameter extraction function" that, when applied to a random variable, will extract the required transform parameters.

In other words, transform objects are no longer random variable _instance_-specific, but random variable _class_-specfic.

brandonwillard committed:
Make transform objects stateless

michaelosthege commented on pymc3/distributions/__init__.py:
```diff
@@ -141,7 +143,7 @@ def change_rv_size(
 def rv_log_likelihood_args(
     rv_var: TensorVariable,
-    transformed: Optional[bool] = True,
+    return_observations: Optional[bool] = True,
```
Would it be a good practice to require `return_observations` to be a kwarg?
The rule of thumb I know if is that one should do that if the passed value (True/False) does not imply the function.

brandonwillard replied:
You mean recast `return_observations` as the "don't return observations"?

michaelosthege replied:
No, just to insert a `, *,` such that writing `return_observations=` is required

brandonwillard replied:
Oh, I see; yeah, we can do that.

michaelosthege commented on pymc3/distributions/__init__.py:
```diff
@@ -233,8 +238,9 @@ def sample_to_measure_vars(
-    scaling: Optional[bool] = True,
+    jacobian: bool = True,
+    scaling: bool = True,
+    transformed: bool = True,
```
Same here - inserting `, *,` would make the calling code easier to read/understand.
This function is not directly public API right?

brandonwillard replied:
I do expect that `logpt` would be of interest to people implementing their own PyMC3-based code, so it's probably somewhere between public and private API.

michaelosthege replied:
Okay, sounds like inserting `, *,` wouldn't break too much on the user side, but would give us more flexibility and clear code

michaelosthege commented on pymc3/distributions/__init__.py:
```diff
@@ -345,11 +354,17 @@ def transform_logp(logp_var: TensorVariable, inputs: List[TensorVariable]) -> Te
+
+        if transform is not None and rv_var is None:
+            warnings.warn(
+                f"A transform was found for {measure_var}" " but no corresponding random variable"
```
String is a bit messed up.
More importantly: The sentence is a bit incomplete - no variable corresponding to what?

brandonwillard replied:
That `measure_var` doesn't have a random variable associated with it, so there's really nothing else to print or say.  If anything, this should probably be an error condition.

brandonwillard replied:
It might actually make more sense to associate the transform object with the `rv_var` (i.e. the random variable).  I'll have to think about that.

michaelosthege commented on pymc3/distributions/transforms.py:
```diff
@@ -45,19 +41,26 @@ class Transform:
+        rv_var
+            The random variable being transformed
+        rv_value
+            The parameters required for the transform.
```
`rv_value` doesn't sound very intuitive for something that holds the transform parameters. (I was confused by this above already.)

How about `rv` and `transform_params`?
Or `rv` and `params`?

brandonwillard replied:
We need to be clear about the `rv_var` and `rv_value[_var]` distinctions.

`rv_var`s are the "sample-space" variables that are produced by `RandomVariable` `Op`s.
`rv_value[_var]`s are the "measure-space" (or log-likelihood) variables that correspond to a specific value of an `rv_var`.

These are the same two types of variables described [here](https://github.com/pymc-devs/pymc3/pull/4489#discussion_r593541630), where the sloppy `P(X = x)` or `x ~ X` notation denotes the `rv_var` with `X` (i.e. the random variable), and the value variable with `rv_value[_var]`.

These transform methods are getting _those_ two variables, so any new name that involves "params" would be inaccurate, because the `rv_value` variable does not provide parameters.  The first argument, `rv_var`, _does_ provide access to a random variable's parameters via `rv_var.owner.inputs`, and&mdash;again&mdash;`rv_value` is a value that's compatible with the random variable `rv_var` (i.e. a value that could've been a sample from it).

michaelosthege replied:
So `rv_var` is the tensor of the user-provided, observed values? (A `TensorConstant`?)

We might still want to copy parts of your explanation into the docstring.

michaelosthege commented on pymc3/distributions/transforms.py:
```diff
@@ -128,19 +135,19 @@ def jacobian_det(self, x):
-    def backward(self, x):
-        return aet.nnet.softplus(x)
+    def backward(self, rv_var, rv_value):
+        return aet.nnet.softplus(rv_value)
```
Are the `rv_var` and `rv_value` mistaken for one another?
Before `x` was the tensor that the transformation was applied to. Reading the docstring:
```
        rv_var
            The random variable being transformed
        rv_value
            The parameters required for the transform.
```
I expected this line to read `aet.nnet.softplus(rv_var)`.

If I'm right this is switched around for multiple transforms above and below too. Otherwise please clarify :)

brandonwillard replied:
The docstrings are wrong; they're the result of a few revisions.  I'll fix that.

michaelosthege reviewed (commented):
After reading the entire diff I'm now quite sure I got the purposes of the `rv_var`  and `rv_value` args wrong.

brandonwillard committed:
Add tests for two important open logpt and Model issues

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -553,6 +553,30 @@ def RandomPdMatrix(n):
+        x = pm.Normal("x", 0, 1, observed=obs)
+        pm.Normal("y", x, 1, observed=obs)
+
+    model.logp(model.test_point)
```
What is the test here, that it computes the test point?

brandonwillard replied:
This test is unfinished (and it was essentially copied from somewhere else), but it makes sure that the `Observed`-typed `x` can be used properly as the input to another variable's log-likelihood.

The next commit I'm going to push gets rid of `Observed`, though, so the problem that this test highlights won't matter after that.

michaelosthege commented on pymc3/tests/test_distributions.py:
```diff
@@ -553,6 +553,30 @@ def RandomPdMatrix(n):
+    assert len(ancestors_with_owner) > 0
+    assert not any(isinstance(v.owner.op, RandomVariable) for v in ancestors_with_owner)
+    assert x.tag.value_var in logpt_ancestors
+    assert y.tag.value_var in logpt_ancestors
```
Would this be a good place to mock the respective `transform` function and do an `assert_called_once_with`?

brandonwillard replied:
Are we concerned about the number of times a transform is called?

michaelosthege replied:
I guess not really, I just wanted to suggest a more targeted way to test the behavior.

michaelosthege commented on pymc3/tests/test_transforms.py:
```diff
@@ -85,18 +92,14 @@ def check_jacobian_det(
-    with pytest.warns(
-        DeprecationWarning, match="The argument `eps` is deprecated and will not be used."
-    ):
-        tr.StickBreaking(eps=1e-9)
```
(Where) do we keep a list of these changes? We should mention them in the release notes. The alternative is to `raise` the `DeprecationWarning` which saves users from complicated digging.

michaelosthege reviewed (dismissed):
There are a few threads still open.
Nevertheless I'll say LGTM, but don't count too much on my judgement. Most of my trust comes from the facts that Brandon did this and that the CI Tests are now ✔.

brandonwillard commented:
Sorry, been pretty busy, but I have another commit to push, and it's a big refactor that should address most/all of the open `logp`-related issues.

brandonwillard committed:
Add non_sequences to uses of Scan Op

This make `aesara.graph.basic.clone_replace` work correctly when `Scan`s are
included in a graph.

brandonwillard committed:
Replace Observed Op with tag.observations

brandonwillard committed:
Add missing imports to pymc3.step_methods.gibbs

brandonwillard committed:
Comment out unused moments

brandonwillard committed:
Make logpt work correctly for nested models and transforms

brandonwillard committed:
Disable use of Arviz in pymc3.tests.test_data_container

brandonwillard committed:
Set model seed correctly in pymc3.tests.test_ndarray_backend

michaelosthege commented on pymc3/distributions/__init__.py:
```diff
@@ -161,80 +157,119 @@ def rv_log_likelihood_args(
 
     """
+    if not var.owner:
+        return None, None
```
Doesn't match with the return type hints and docstring.

Can you explain (maybe in the docstring) why and under what circumstances `None, None` is returned?

michaelosthege commented on pymc3/distributions/__init__.py:
```diff
@@ -249,117 +284,96 @@ def logpt(
+        rv_value = rv_var.type.filter_variable(rv_value.astype(rv_var.dtype))
+
+        if rv_value_var is None:
+            rv_value_var = rv_value
```
That's the case when `rv_value` has no observations, right?

michaelosthege commented on pymc3/distributions/continuous.py:
```diff
@@ -1160,8 +1153,8 @@ def dist(cls, alpha=None, beta=None, mu=None, sigma=None, sd=None, *args, **kwar
-        mean = alpha / (alpha + beta)
-        variance = (alpha * beta) / ((alpha + beta) ** 2 * (alpha + beta + 1))
+        # mean = alpha / (alpha + beta)
+        # variance = (alpha * beta) / ((alpha + beta) ** 2 * (alpha + beta + 1))
```
can be removed?

michaelosthege commented on pymc3/distributions/distribution.py:
```diff
@@ -111,12 +111,12 @@ def logp(op, value, *dist_params, **kwargs):
+            #
+            #     @logp_transform.register(rv_type)
+            #     def transform(op, *args, **kwargs):
+            #         return class_transform(*args, **kwargs)
```
TODO

michaelosthege commented on pymc3/distributions/distribution.py:
```diff
@@ -328,26 +328,17 @@ def _distr_parameters_for_repr(self):
-        super().__init__(shape, dtype, defaults=defaults, *args, **kwargs)
+        if kwargs.get("transform", None):
+            raise ValueError("Transformations for discrete distributions")
+
```
Shouldn't we keep the `dtype` checks? (Based on `intX`.)

brandonwillard replied:
Those are done at the Aesara `Op`-level now (i.e. within `RandomVariable.make_node`); although I'm not sure if float-to-int conversion is part of that.  It might only raise an exception for the wrong dtype.  If it's not, then we might need to add that at this level.

michaelosthege commented:
The failing test looks like the non-deterministic `logpt` that @ricardoV94 noticed a few days ago?

brandonwillard committed:
Prevent Model from turning on test value computations
